### DISCUSSION TITLE
ConversableAgent class needs much love

### DISCUSSION THREAD

**Discussion by bitnom** [microsoft/autogen#1247] [Ideas]
What do we think about [conversable_agent.py](https://github.com/microsoft/autogen/blob/main/autogen/agentchat/conversable_agent.py) ?

It feels like it's well on its way to becoming a tangled web of tears. There are so many patterns happening. I've been wanting to refactor it but don't want to bother unless maintainers see the Matrix with me.. I feel like we could do without a lot of what's going on here and implement:

- some nice chainable observers.
- separation of some of these concerns
- single responsibility principal
- use pydantic for most the things
- replace these one-line class property methods
- deduplicate all the async
- no more what seems like factory, command, and strategy patterns altogether
- maybe more composition. Maybe not though. Probably leave most of the subclassing alone until last or later.

To be clear, I'm not hating on the code. Technical debt is what we do. I did something similar myself recently with a command factory that used dependency injection, and though it technically worked, I had to get rid of it because it was just too cool.

I'm happy to be wrong if I am.

**Comment by ekzhu** [comment]
You are welcome join the discussion in #1240. We have some idea there.
